architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+  
+### User scenarios (optional)
+
+The scenario document for user-defined types is in progress [here](https://github.com/radius-project/design-notes/pull/58). Please refer to the scenarios defined in that doc.
+
+## User Experience (if applicable)
+
+Proposed user-experience changes are defined in the [scenario doc](https://github.com/radius-project/design-notes/pull/58).
+
+Here's a abstract summary. 
+
+Jenny is a platform engineer implementing support for PostgreSQL in Radius. Her E2E experience authoring a declarative user-defined type looks like this:
+
+**Authoring**
+
+1. Choose a name for the new type. She chooses `Contoso.Example/postgreSQLDatabases`.
+2. Design a schema for the new type. She's inspired by the design of `Applications.Datastores/sqlDatabases` (MS SQL Server) and chooses something similar. Her new type will return fields like `database`, `port`, `host`, `username`, `password`, and `url`. 
+3. Author a manifest for the new resource provider `Contoso.Example`. She needs to make a few choices at this stage:
+   1. Choose an API version, she picks `2024-08-1`. 
+   2. Implement an RP imperatively in code, or declaratively with support for recipes. She wants to use recipes and the declarative approach.
+4. Use the tools + manifest to generate:
+   1.  A Bicep extension (like the `radius` or `aws` Bicep extension)
+   2.  A "cooked" version of the manifest that can be used with `rad resourceprovider create ...`. 
+
+*At this point she's done authoring code/assets for the new type and resource provider.*
+
+**Using**
+
+1. Use `rad resourceprovider create ...` to register the resource provider with Radius.
+2. Use `rad recipe register ...` to register a recipe for `Contoso.Example/postgreSQLDatabases`
+3. Add the Bicep extension to her local `bicepconfig.json`. Now she can add `extension contoso` to a Bicep file and use `Contoso.Example/postgreSQLDatabases` from Bicep.
+4. `rad deploy app.bicep`
+
+**Publishing (optional)**
+
+*Depending on who she wants to share her work with, she might do a few different things.*
+
+1. Publishing to the internet.
+   1. She publishes the Bicep provider to a public OCI registry.
+   2. She hosts the provider manifest on `github.com` in a personal public repo.
+   3. She publishes instructions for use and some example recipes in the same public repo.
+2. Publishing internally to her company.
+   1. She publishes the Bicep provider to a public OCI registry.
+   2. She adds the provider manifest to internal repos that are used with GitOps. Now the type is available in all of their internal Radius installations.
+   3. She does the same with recipes.
+   4. She adds instructions for use to internal wikis and templates.
+3. Contributing to upstream
+   1. She forks the `radius-project/types` repo (real name TBD).
+   2. She follows an issue template in the repo and creates a pull-request to make her new type available as `Applications.Datastores/postgreSQLDatabases`.
```

Comment:
Why `Applications.Datastores/postgreSQLDatabases`? Why not `Contoso.Example/postgreSQLDatabases`? Are we changing the RPs when we are moving the new type to the official repository?

Reply:
I'm making a distinction here about the official project. `Applications.*` namespaces are reserved for Radius.